File: src/main/java/de/tectoast/emolga/commands/flegmon/NextBirthdayCommand.java
package de.tectoast.emolga.commands.flegmon;

import de.tectoast.emolga.commands.GuildCommandEvent;
import de.tectoast.emolga.commands.PepeCommand;
import de.tectoast.emolga.utils.sql.DBManagers;
import de.tectoast.emolga.utils.sql.managers.BirthdayManager;
import net.dv8tion.jda.api.entities.TextChannel;

import java.sql.SQLException;
import java.util.Calendar;
import java.util.Comparator;
import java.util.HashMap;
import java.util.Map;

public class NextBirthdayCommand extends PepeCommand {
    public NextBirthdayCommand() {
        super("nextbirthday", "Zeigt die naheliegende Geburtstage an");
        setArgumentTemplate(ArgumentManagerTemplate.noArgs());
        aliases.add("nextbirthdays");
    }

    @Override
    public void process(GuildCommandEvent e) throws SQLException {
        TextChannel tco = e.getChannel();
        Calendar curr = Calendar.getInstance();
        curr.set(Calendar.HOUR_OF_DAY, 0);
        curr.set(Calendar.MINUTE, 0);
        curr.set(Calendar.SECOND, 0);
        Map<Long, Calendar> map = new HashMap<>();
        for (BirthdayManager.Data bData : DBManagers.BIRTHDAYS.getAll()) {
            Calendar c = Calendar.getInstance();
            c.set(Calendar.DAY_OF_MONTH, bData.day());
            c.set(Calendar.MONTH, bData.month() - 1);
            c.set(Calendar.YEAR, curr.getTimeInMillis() - c.getTimeInMillis() >= 0 ? c.get(Calendar.YEAR) + 1 : c.get(Calendar.YEAR));
            c.set(Calendar.HOUR_OF_DAY, 0);
            c.set(Calendar.MINUTE, 0);
            c.set(Calendar.SECOND, 0);
            long dif = c.getTimeInMillis() - curr.getTimeInMillis();
            if (dif <= 1209600000/* && dif >= -432000000*/) {
                map.put(bData.userid(), c);
            }
        }
        if (map.isEmpty()) {
            tco.sendMessage("Es gibt keine nahegelegenen Geburtstage!").queue();
            return;
        }
        StringBuilder str = new StringBuilder("Die nächsten Geburtstage:\n\n");
        HashMap<Long, String> names = new HashMap<>();
        e.getGuild().retrieveMembersByIds(map.keySet()).get().forEach(mem -> names.put(mem.getIdLong(), mem.getEffectiveName()));
        for (long id : map.keySet().stream().sorted(Comparator.comparing(s -> map.get(s).getTimeInMillis())).toList()) {
            Calendar c = map.get(id);
            str.append("`").append(getWithZeros(c.get(Calendar.DAY_OF_MONTH), 2)).append(".").append(getWithZeros(c.get(Calendar.MONTH) + 1, 2)).append(".").append("`: ").append(names.get(id)).append("\n");
        }
        tco.sendMessage(str.toString()).queue();
            /*(s1, s2) -> {
            Calendar c1 = map.get(s1);
            Calendar c2 = map.get(s2);
            return Long.compare(c1.getTimeInMillis(), c2.getTimeInMillis());
        }*/
    }
}

File: src/main/kotlin/de/tectoast/emolga/commands/flegmon/NextBirthdayCommand.kt
package de.tectoast.emolga.commands.flegmon

import de.tectoast.emolga.commands.GuildCommandEvent
import de.tectoast.emolga.commands.PepeCommand
import de.tectoast.emolga.utils.sql.managers.BirthdayManager
import net.dv8tion.jda.api.entities.Member
import java.util.*
import java.util.function.Consumer

class NextBirthdayCommand : PepeCommand("nextbirthday", "Zeigt die naheliegende Geburtstage an") {
    init {
        argumentTemplate = ArgumentManagerTemplate.noArgs()
        aliases.add("nextbirthdays")
    }

    override fun process(e: GuildCommandEvent) {
        val tco = e.textChannel
        val curr = Calendar.getInstance()
        curr[Calendar.HOUR_OF_DAY] = 0
        curr[Calendar.MINUTE] = 0
        curr[Calendar.SECOND] = 0
        val map: MutableMap<Long, Calendar> = HashMap()
        for (bData in BirthdayManager.all) {
            val c = Calendar.getInstance()
            c[Calendar.DAY_OF_MONTH] = bData.day
            c[Calendar.MONTH] = bData.month - 1
            c[Calendar.YEAR] = if (curr.timeInMillis - c.timeInMillis >= 0) c[Calendar.YEAR] + 1 else c[Calendar.YEAR]
            c[Calendar.HOUR_OF_DAY] = 0
            c[Calendar.MINUTE] = 0
            c[Calendar.SECOND] = 0
            val dif = c.timeInMillis - curr.timeInMillis
            if (dif <= 1209600000 /* && dif >= -432000000*/) {
                map[bData.userid] = c
            }
        }
        if (map.isEmpty()) {
            tco.sendMessage("Es gibt keine nahegelegenen Geburtstage!").queue()
            return
        }
        val str = StringBuilder("Die nächsten Geburtstage:\n\n")
        val names = HashMap<Long, String>()
        e.guild.retrieveMembersByIds(map.keys).get()
            .forEach(Consumer { mem: Member -> names[mem.idLong] = mem.effectiveName })
        for (id in map.keys.stream().sorted(Comparator.comparing { s: Long ->
            map[s]!!
                .timeInMillis
        }).toList()) {
            val c = map[id]
            str.append("`").append(getWithZeros(c!![Calendar.DAY_OF_MONTH], 2)).append(".").append(
                getWithZeros(
                    c[Calendar.MONTH] + 1, 2
                )
            ).append(".").append("`: ").append(names[id]).append("\n")
        }
        tco.sendMessage(str.toString()).queue()
        /*(s1, s2) -> {
            Calendar c1 = map.get(s1);
            Calendar c2 = map.get(s2);
            return Long.compare(c1.getTimeInMillis(), c2.getTimeInMillis());
        }*/
    }
}